test/custom-actions-spec.js

```diff
@@ -0,0 +1,195 @@
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('executes a custom action when a document is deleted', function() {
+      var oldDoc = { _id: docType };
+
+      testHelper.verifyDocumentDeleted(oldDoc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('does not execute a custom action if authorization was denied', function() {
+      var doc = { _id: docType };
+
+      testHelper.verifyAccessDenied(doc, null, [ 'write' ]);
+      verifyCustomActionNotExecuted();
+    });
+  });
+
+  describe('the onValidationSucceeded event', function() {
+    var docType = 'onValidationDoc';
+
+    it('executes a custom action when a document is created', function() {
+      var doc = { _id: docType };
+
+      testHelper.verifyDocumentCreated(doc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('executes a custom action when a document is replaced', function() {
+      var doc = { _id: docType };
+      var oldDoc = { _id: docType };
+
+      testHelper.verifyDocumentReplaced(doc, oldDoc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('executes a custom action when a document is deleted', function() {
+      var oldDoc = { _id: docType };
+
+      testHelper.verifyDocumentDeleted(oldDoc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('does not execute a custom action if the document contents are invalid', function() {
+      var doc = {
+        _id: docType,
+        unsupportedProperty: 'foobar'
+      };
+
+      testHelper.verifyDocumentNotCreated(doc, docType, errorFormatter.unsupportedProperty('unsupportedProperty'));
+      verifyCustomActionNotExecuted();
+    });
+  });
+
+  describe('the onAccessAssignmentsSucceeded event', function() {
+    var docType = 'onAccessAssignmentsDoc';
+
+    it('executes a custom action when a document is created', function() {
+      var doc = { _id: docType };
+
+      testHelper.verifyDocumentCreated(doc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('executes a custom action when a document is replaced', function() {
+      var doc = { _id: docType };
+      var oldDoc = { _id: docType };
+
+      testHelper.verifyDocumentReplaced(doc, oldDoc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('executes a custom action when a document is deleted', function() {
+      var oldDoc = { _id: docType };
+
+      testHelper.verifyDocumentDeleted(oldDoc);
+      verifyCustomActionExecuted(docType);
+    });
+
+    it('does not execute a custom action if the document definition does not define access assignments', function() {
+      var doc = { _id: 'missingAccessAssignmentsDoc' };
+
+      testHelper.verifyDocumentCreated(doc);
+      verifyCustomActionNotExecuted();
+    });
+  });
+
+  describe('the onDocumentChannelAssignmentSucceeded event', function() {
+    var docType = 'onDocChannelsAssignedDoc';
+
+    it('executes a custom action when a document is created', function() {
```

Comment:
Again, if you wanted to be clever, you could pull these three cases out into a reusable harness to be shared across all custom action doc type suites